### DISCUSSION TITLE
Breezy Weather - 1 month since first release!

### DISCUSSION THREAD

**Discussion by papjul** [breezy-weather/breezy-weather#226] [Announcements]
It's been about a month since the first (alpha) release of Breezy Weather, it’s time to take a look back at what was done since then! If you had a hard time following all changelogs, this is the one thing to read ;)

# Main highlights

- Translation can now be done [from Weblate](https://hosted.weblate.org/projects/breezy-weather/breezy-weather-android/#information), making translation process easier for everyone. Many languages were added or updated since then, and we now have 30% more translations compared to Geometric Weather, a huge thanks to all contributors!
- Settings are reorganized making things easier to find
- Onboarding was reviewed: user is asked to add its location with GPS or manually with search, instead of being prompted GPS permissions right away. A weather source selection dialog is also available for current location instead of choosing AccuWeather by default.
- Add details about pollutants when tapping on them.
- Alerts are now more visible and show their color if feature available from weather source.
- Tons of fixes and small improvements

Weather sources-specific:
- **[All]** More data is now computed or extrapolated when missing, making some weather sources much more complete
- **[AccuWeather]** Add air quality
- **[Open-Meteo]** Add air quality
- **[MET No]** Add Nowcast for Nordic area
- **[MET No]** Add air quality for Norway
- **[China]** Search now possible with English names of Chinese cities
- **[China]** Precipitation in next hour

V4.4.0-beta containing all changes mentioned above is now available as pre-release, and will be pushed to IzzyDroid on Thursday if no major regression is reported.

# Technical

Technical was the main focus of the past month, although it was probably transparent to you.

Breezy Weather now has about 25 % less code* than Geometric Weather (1.93 MB, down from 2.57) while having more features and improvements.

*Not the same thing as APK size which keeps growing, especially as translations gets completed

To achieve this:
1) I rewrote all Java code in **Kotlin** in v4.1.0, making the code mode readable at the same time.
2) I removed a lot of spaghetti and duplicated code (although there is still some), making it much easier to maintain.
3) Some components were rewritten in Jetpack Compose, making it less verbose than, for example, RecyclerView.
4) In v4.2.0, I rewrote the background update logic to make it more up to date with Android standards. This rewrite helped fix many issues and make it more battery-friendly. Note: if you no longer have background updates since this version, please [check the FAQ](https://github.com/breezy-weather/breezy-weather/blob/main/HELP.md).
5) In v4.4.0, I rewrote the "complete missing data from weather sources" logic, reducing a lot of code reused in many sources, while adding support for more data at the same time! This change also makes implementing new weather sources at everyone's reach.

Such a big rewrite doesn't come without regressions, but I'm rather impressed by the low ratio of bugs per lines of code modified (especially, given there is almost no automated tests). Thanks to all users who reported these issues to GitHub, with a special mention to min7 who reported many and helped investigating.

Most of them are now fixed, and we are down to 20 bugs currently opened:
- 10 were already present in Geometric Weather
- 2 are waiting for more info (they may not be actual bugs)
- 8 other are either cosmetic or have workaround

# What's next

Major rewritings are mainly done, many bugs are fixed, so we are heading towards getting the app more and more stable.

There are still a couple things to get it out of beta and on F-Droid for now.

If I'm not encountering any pitfalls, I believe the next major update will allow **per-location settings**, especially allowing to **complete data with another source** (for example, you could have a location added via OpenWeather completed with Air quality data from Open-Meteo, all in the same location, so you don’t need to add the city twice and switch between both).

On the technical part, there is still a **major rewrite to be done on the main screen**, as the code goes a bit in all ways and it becomes hard to fix issues and make improvements there.

Along the way, I will of course bring more QoL improvements and keep merging translation updates.

# Final note

Finally, throwing a bottle into the sea: if any experienced Android developer is coming by reading this and would like to help, even for just a few hours, it would be appreciated! You can join this Matrix channel for discussion: `#breezy-weather-developer-discussion:matrix.org`.

Things I could need help with, depending on what you like/time available:
- code review
- architecture review
- advice on possible implementation of feature X
- pull requests for issues opened to contributors
- unit tests
- etc.

Until then, I will continue my work on the project, with no promises on any schedule though!

**Comment by gbakeman** [comment]
Excellent work, thank you.

**Comment by bananabr34d** [comment]
I've been using Geometric Weather for a long time and had no idea that it had been forked.  I am so glad I found this and want to say thank you so much for all your great work.
